Pure Javascript Drafter

danielgtaylor opened the pull request:
This builds on top of the work from #22. Summary:

* Rebase on `master` and default to Refract output
* Uses Docker to build `jsdrafter.js`
* Runs the build on CircleCI
* Exposes a single `parse` function that takes options similar to the `protagonist.parseSync` function in Protagonist.

##### Example usage

Node.js:

```js
var drafter = require('./jsdrafter.js');
var result = drafter.parse('# Your API Blueprint');
```

Browser:

```html
<script src="./jsdrafter.js"></script>
<script>
  var result = drafter.parse('# Your API Blueprint');
</script>
```

w-vi commented:
@danielgtaylor  this code looks much better than mine :)

kylef commented on README.md:
````diff
@@ -106,40 +106,89 @@ See [parse feature](features/parse.feature) for the details on using the `drafte
+
+2. Build
+    ```sh
+    $ docker pull "apiaryio/base-emscripten-dev"
````
I can't find this, is it private?

danielgtaylor replied:
It's from here: https://hub.docker.com/r/apiaryio/base-emscripten-dev/

kylef replied:
Ah thanks, proves what I know about docker. Assumed `apiaryio/base-emscripten-dev` was a git repo on GitHub.

kylef commented:
Further idea, so people don't have to build this themselves. We could use TravisCI/CircleCI release support to upload the built .js file to the releases.

- https://github.com/apiaryio/Paw-APIBlueprintImporter/blob/master/.travis.yml#L10-L19
- https://github.com/apiaryio/Paw-APIBlueprintImporter/releases/tag/1.1.1

pksunkara commented on emcc/test-jsdrafter:
```diff
@@ -0,0 +1,62 @@
+        var p = path.basename(file, '.apib');
+        try {
+            if (fs.statSync(path.join(fixtures, p+".result.json")).isFile()) {
+                apibs.push({apib:path.join(fixtures,file),
```
Can you properly indent this object?

pksunkara commented on emcc/test-jsdrafter:
```diff
@@ -0,0 +1,62 @@
+                            result:path.join(fixtures, p+".result.json")});
+                testRun.total++;
+            } else {
+                console.error("Found "+path.join(fixtures,file)+" but not result!");
```
Please leave spaces before and after `+`. Similarly, on line 17 of this file.

pksunkara commented on emcc/test.html:
```diff
@@ -0,0 +1,50 @@
+    Hello World.
+
+</textarea>
+  <div id=output>XXX</div>
```
Shouldn't we put `OK` by default here?

pksunkara commented:
lgtm. Just a few minor things. Also, I agree with kyle that we may want to setup Travis to do automatic releases whenever a new version is tagged in C++ drafter.

pksunkara commented:
By the way, why are we still sticking with the name `jsdrafter.js`? We can use `drafter.js` simply.

kylef commented on emcc/test-jsdrafter:
```diff
@@ -0,0 +1,62 @@
+var apibs = [];
+fs.readdirSync(fixtures).forEach(function(file) {
+    if (path.extname(file) === ".apib") {
+        var p = path.basename(file, '.apib');
```
Minor nit, changing between `"` and `'` on line/line above.

danielgtaylor commented:
I have updated the code with the following changes:

* Put `uglify` information back into the README
* Renamed `jsdrafter.js` to `drafter.js`
* Generate a `.mem` file by default, and a `nomem` version as well.
* Create an `npm` module `drafter.js` (not `drafter`) which will use the `nomem` version.
* Update the `drafter.js` test script to test all fixtures, with a blacklist to ignore some apparently outdated / unused ones.
* Create a release script for CircleCI to push artifacts to GitHub and `npm`
* Consistency fixes (e.g. type of quotes used)
* Default HTML example to `OK` rather than `XXX`

danielgtaylor commented:
@pksunkara I've completely rewritten the js tests and now everything passes:

* Compare output of `drafter` executable to `drafter.js` output
  * Tests both `ast` and `refract` output types
* Skips any file where the `drafter` executable returns nonzero
* Prints timing information

It takes about three minutes to complete the test on CircleCI now. It could be sped up a little but the tests are done in series to get accurate timing information. Looks like you can expect a 10x slowdown for relatively small API Blueprint files when using the pure JS parser. Results:

```
Test run result:
================
Total:  172
Skipped: 6
Passed: 166
Failed: 0
```

pksunkara commented on README.md:
````diff
@@ -106,40 +106,104 @@ See [parse feature](features/parse.feature) for the details on using the `drafte
+    $ drafter --help
+    ```
+
+## drafter.js
````
I think we should move this whole section to be after `# Bindings` because we want people to use native bindings first and then if they are not working go to already compiled versions.

pksunkara commented:
> Skips any file where the drafter executable returns nonzero

Why is that so? Are all of them not rendering any JSON object at all? Can I get the names of those files to verify once?

danielgtaylor commented:
@pksunkara the idea was that for a failure we don't really care about the output (for warnings we do). Given that they pass and this was controversial I've gotten rid of the skipping feature and it now compares the output for those as well. I've also started tracking the timing info in the summary:

```
Test run result:
================
Total:  172
Passed: 172
Failed: 0
Average JS speed: 73.4 times slower than C++
```

README has also been updated and this should be ready to be merged!
